src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java

```diff
@@ -107,29 +109,47 @@ public static boolean isCigarValid(Cigar cigar) {
         return false;
     }
 
-    public static final int countRefBasesBasedOnCigar(final SAMRecord read, final int cigarStartIndex, final int cigarEndIndex){
+    /**
+     * Compute the number of reference bases between the start (inclusive) and end (exclusive) cigar elements.
+     */
+    @SuppressWarnings("fallthru")
+    public static int countRefBasesBasedOnCigar(final SAMRecord read, final int cigarStartIndex, final int cigarEndIndex){
+        if (read == null){
+            throw new IllegalArgumentException("null read");
+        }
+        final List<CigarElement> elems = read.getCigar().getCigarElements();
+        if (cigarStartIndex < 0 || cigarEndIndex > elems.size() || cigarStartIndex > cigarEndIndex){
+            throw new IllegalArgumentException("invalid index:" + 0 + " -" + elems.size());
+        }
         int result = 0;
-        for(int i = cigarStartIndex; i<cigarEndIndex;i++){
-            final CigarElement cigarElement = read.getCigar().getCigarElement(i);
+        for(int i = cigarStartIndex; i < cigarEndIndex; i++){
+            final CigarElement cigarElement = elems.get(i);
             switch (cigarElement.getOperator()) {
                 case M:
-                case S:
                 case D:
                 case N:
-                case H:
+                case EQ:
+                case X:
+                case S:        //TODO is this correct? CigarOperator.consumesReferenceBases is false
+                case P:        //TODO is this correct? CigarOperator.consumesReferenceBases is false
```

Comment:
I'm not sure how P is considered, is it counted in the reference as a base, for example if I have read starting at position 10 and it cigar is 1M3P1M does the last base facing position 14 or 11? if it is the first case, then it is fine, if it is the second, then it should be treated like I